crates/ruma-events/src/room/join_rules.rs

```diff
@@ -25,67 +23,130 @@ pub type JoinRulesEvent = StateEvent<JoinRulesEventContent>;
 pub struct JoinRulesEventContent {
     /// The type of rules used for users wishing to join this room.
     #[ruma_event(skip_redaction)]
+    #[serde(flatten)]
     pub join_rule: JoinRule,
-
-    /// Allow rules used for the `restricted` join rule.
-    #[cfg(feature = "unstable-pre-spec")]
-    #[serde(default)]
-    #[ruma_event(skip_redaction)]
-    pub allow: Vec<AllowRule>,
 }
 
 impl JoinRulesEventContent {
     /// Creates a new `JoinRulesEventContent` with the given rule.
     pub fn new(join_rule: JoinRule) -> Self {
-        Self {
-            join_rule,
-            #[cfg(feature = "unstable-pre-spec")]
-            allow: Vec::new(),
-        }
+        Self { join_rule }
     }
 
     /// Creates a new `JoinRulesEventContent` with the restricted rule and the given set of allow
     /// rules.
     #[cfg(feature = "unstable-pre-spec")]
     pub fn restricted(allow: Vec<AllowRule>) -> Self {
-        Self { join_rule: JoinRule::Restricted, allow }
+        Self { join_rule: JoinRule::Restricted(Restricted::new(allow)) }
     }
 }
 
 /// The rule used for users wishing to join this room.
 ///
 /// This type can hold an arbitrary string. To check for formats that are not available as a
 /// documented variant here, use its string representation, obtained through `.as_str()`.
-#[derive(Clone, Debug, PartialEq, Eq, StringEnum)]
-#[ruma_enum(rename_all = "lowercase")]
-#[non_exhaustive]
+#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[serde(tag = "join_rule")]
 pub enum JoinRule {
     /// A user who wishes to join the room must first receive an invite to the room from someone
     /// already inside of the room.
+    #[serde(rename = "invite")]
     Invite,
 
     /// Reserved but not yet implemented by the Matrix specification.
+    #[serde(rename = "knock")]
     Knock,
 
     /// Reserved but not yet implemented by the Matrix specification.
+    #[serde(rename = "private")]
     Private,
 
     /// Users can join the room if they are invited, or if they meet any of the conditions
     /// described in a set of [`AllowRule`]s.
     #[cfg(feature = "unstable-pre-spec")]
-    Restricted,
+    #[serde(rename = "restricted")]
+    Restricted(Restricted),
 
     /// Anyone can join the room without any prior action.
+    #[serde(rename = "public")]
     Public,
 
     #[doc(hidden)]
-    _Custom(String),
+    _Custom(CustomJoinRule),
 }
 
 impl JoinRule {
     /// Creates a string slice from this `JoinRule`.
```

Comment:
This doc comment should be updated to clarify that this just returns the `join_rules` string, now that this type stores additional things as well.